### PR TITLE
Initial commit of Aleph provider with working login. BBS-13

### PR DESCRIPTION
**Pull request by tripox**

### CODE REVIEW

File: modules/aleph/aleph.install [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+  if (!function_exists('simplexml_load_string')) {
+    $requirements['simplexml'] = array(
+      'title' => 'SimpleXML',
+      'description' => $t('The Aleph module requires SimpleXML to function. Please install and/or enable SimpleXML in your PHP configuration.'),
+      'severity' => REQUIREMENT_ERROR,
+    );
+  }
+
+  if ($phase == 'runtime') {
```
Comment (danquah) [inline modules/aleph/aleph.install:23]: Have you tested that the site can still be installed and that it is possible to select the aleph provider during installation?

File: modules/aleph/aleph.info [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+dependencies[] = ding_reservation
+dependencies[] = ding_user
+dependencies[] = email
+dependencies[] = entity
+dependencies[] = features
+dependencies[] = list
+dependencies[] = options
+dependencies[] = profile2
+dependencies[] = text
+dependencies[] = virtual_field
```
Comment (danquah) [inline modules/aleph/aleph.info:17]: I suspect that we have dependencies on modules we dont need right now on this list, but lets check it when we get a bit further and know what we actually need.

File: modules/aleph/aleph.module [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+ */
+function aleph_client() {
+  // This is basically a singleton. We also wait until right before
+  // instantiating to include our required classes. That seems like a
+  // decent performance trade-off instead of loading everything on every
+  // page load.
+  static $client;
+  if (!isset($client)) {
+    try {
+      $client = new AlephClient(variable_get('aleph_base_url', 'http://snorri.lb.is/X'));
```
Comment (danquah) [inline modules/aleph/aleph.module]: Think we should assume the variable is set, and fail if it is not - the variable can be set during installation, and for development-purposes we could set it in docker.settings.php

File: modules/aleph/includes/aleph.user_consent.inc [hunk 4]
```diff
@@ -0,0 +1,86 @@
+<?php
+/**
+ * @file
+ * Implements the Aleph user consent interface.
+ *
+ * When a user gives permission to store his or her loan history for longer
```
Comment (danquah) [inline modules/aleph/includes/aleph.user_consent.inc:7]: The documentation should reflect that we hardcode the consent on purpose.

File: modules/aleph/includes/aleph.user_consent.inc [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    'loan_history_store' => array(
+      'title' => t('Personalisation'),
+      'description' => t('In order to use this function, we request your permission to store your personal information'),
+      'required' => TRUE,
+    ),
+  );
+}
+
+/**
+ * Update the user consent.
```
Comment (danquah) [inline modules/aleph/includes/aleph.user_consent.inc]: We should probably also document pr. method why we hardcode the response

File: modules/aleph/lib/AlephClient/AlephClient.class.php [hunk 6]
```diff
@@ -0,0 +1,164 @@
+<?php
+/**
+ * @file
+ * Provides a client for the Axiell Aleph library information webservice.
+ */
+
+require __DIR__ . '../../../vendor/autoload.php';
```
Comment (danquah) [inline modules/aleph/lib/AlephClient/AlephClient.class.php:7]: We need to be 100% on how we autoload classes if we have more than one than one autoload.php in play (we might be good, and might be in trouble)

**Review by danquah — CHANGES_REQUESTED** [review]

File: modules/aleph/lib/AlephClient/AlephClient.class.php (hunk 6 above)
Comment (tripox) [reply modules/aleph/lib/AlephClient/AlephClient.class.php:7]:
```shell
grep -r 'guzzle' .
```
in the root shows that guzzle is used in two modules and then theres a library (I removed some of the output for convenience):

```shell
./web/profiles/ding2/modules/fbs/composer.json:        "guzzlehttp/guzzle": "5.2.0"
./web/profiles/ding2/modules/aleph/composer.json:        "guzzlehttp/guzzle": "^6.2"
./web/profiles/ding2/modules/bpi/lib/bpi-client/composer.json:        "guzzlehttp/guzzle": "^6.2"
./web/profiles/ding2/ding2.make:libraries[guzzle][download][tag] = "6.2.2"
./web/profiles/ding2/ding2.make:libraries[guzzle][destination] = "libraries"
./web/profiles/ding2/ding2.make:libraries[promises][download][url] = "https://github.com/guzzle/promises.git"
./web/profiles/ding2/ding2.make:libraries[psr7][download][url] = "https://github.com/guzzle/psr7.git"
```

File: modules/aleph/aleph.info (hunk 2 above)
Comment (tripox) [reply modules/aleph/aleph.info:17]: I agree. :)

File: modules/aleph/aleph.module (hunk 3 above)
Comment (tripox) [reply modules/aleph/aleph.module]: I agree. I've set the variable in `docker.settings.php`: https://github.com/reload/ding-docker/commit/100cc7c240750ae7c15ac919b6ba692cb732ed07 and removed the default variable here: https://github.com/rvk-utd/ding2/pull/1/commits/895819b7d4556ea467be67ff8efc3dac36af1261.

File: modules/aleph/includes/aleph.user_consent.inc (hunk 4 above)
Comment (tripox) [reply modules/aleph/includes/aleph.user_consent.inc:7]: https://github.com/rvk-utd/ding2/pull/1/commits/8aeb4f0c3635aa0042b3e4100c2a74ee829fc88a

File: modules/aleph/includes/aleph.user_consent.inc (hunk 5 above)
Comment (tripox) [reply modules/aleph/includes/aleph.user_consent.inc]: https://github.com/rvk-utd/ding2/pull/1/commits/8aeb4f0c3635aa0042b3e4100c2a74ee829fc88a

File: modules/aleph/aleph.install (hunk 1 above)
Comment (tripox) [reply modules/aleph/aleph.install:23]: I will try now.

**Comment by tripox** [comment]
@danquah, I'll nudge you, when I've tried to run the installation.

File: modules/aleph/aleph.install (hunk 1 above)
Comment (tripox) [reply modules/aleph/aleph.install:23]: Works perfectly. Added patch to show provider when installing profile. https://github.com/rvk-utd/ding2/pull/1/commits/70d7fc82b21ad7fd4aac43fc4faf99396d5b37fd

File: modules/aleph/lib/AlephClient/AlephClient.class.php (hunk 6 above)
Comment (danquah) [reply modules/aleph/lib/AlephClient/AlephClient.class.php:7]: 👍

File: modules/aleph/includes/aleph.user_consent.inc (hunk 5 above)
Comment (danquah) [reply modules/aleph/includes/aleph.user_consent.inc]: Would be nice with a comment pr function that makes a static return.

File: modules/aleph/includes/aleph.user_consent.inc (hunk 4 above)
Comment (danquah) [reply modules/aleph/includes/aleph.user_consent.inc:7]: 👍

File: modules/aleph/aleph.info (hunk 2 above)
Comment (danquah) [reply modules/aleph/aleph.info:17]: 👍

File: modules/aleph/aleph.install (hunk 1 above)
Comment (danquah) [reply modules/aleph/aleph.install:23]: 👍

**Review by danquah — APPROVED** [review]

**Pull request merged**